### DISCUSSION TITLE
Reference error when `x-bind` parameters within `x-for`

### DISCUSSION THREAD

**Discussion by GamesmenJordan** [alpinejs/alpine#2067] [7. Bugs (archived)]
**Description**
When the parameters of an `x-for` loop are used within the context of an `x-bind` on elements within that loop, Alpine will throw a Reference Error for the parameter. This does not impact rendering each of the elements on the DOM and each element does appear to correctly have access to the parameter for other purposes (eg. `x-text` will display the string value of the parameter). But the Reference Error is uncaught and may break the execution of it or other scripts.

This issue does not happen on 2.x version of Alpine following the same setup.

**Example**
2.3.5 : https://jsfiddle.net/netjordanlee/2j71odwb/13/
3.3.3 : https://jsfiddle.net/netjordanlee/h632yxuz/9/

Both use the same Alpine data object:

```
function testComponent() {
	return {
        ready: false,
        myProperty: 'hello',
        myArray: [],
        init() {
            this.myArray = [
            {a:'1a', b:'1b'},
            {a:'2a', b:'2b'},
            {a:'3a', b:'3b'},
          ]

          this.ready = true;
        }
	}
}
```

And similar HTML (removed `x-init` because of the inferred call to `x-init` in 3.x)
```
<div x-data="testComponent()"
     x-init="init()">
     <span x-text="myProperty"></span>
  <template x-for="(element, index) in myArray" :key="element.a">
    <div :class="{ 'test' : (element.b === '2b') }">
      <span x-text="element.a"></span>
      <span x-text="element.b"></span>
    </div>
  </template>
</div>
```

In the console (use DevTools, not the JSFiddle Console) the following error appears in the 3.x version:
```
Uncaught (in promise) ReferenceError: element is not defined
    at eval (eval at generateFunctionFromString (cdn.js:364), <anonymous>:3:32)
    at cdn.js:375
    at tryCatch (cdn.js:399)
    at cdn.js:2241
    at reactiveEffect (cdn.js:901)
    at Object.effect2 [as effect] (cdn.js:876)
    at effect (cdn.js:44)
    at wrappedEffect (cdn.js:60)
    at Function.<anonymous> (cdn.js:2240)
    at flushHandlers (cdn.js:442)
```

Error points to the evaluation of `:class="{ 'test' : (element.b === '2b') }"` being the source of the error, presented as javascript below:
```
(async function anonymous(__self,scope
) {
with (scope) { __self.result = element.b }; __self.finished = true; return __self.result;
})
```

**Comment by SimoTod** [comment]
the issue seems to only happen if the component has a init() function or x-init. Is it happening the same to you?

**Reply by GamesmenJordan** [reply]
I opened this as a bug, might be better to talk there. But after checking, it seems it is related to having an init, all of my alpine data objects have init so I didn't notice it doesn't error without.

**Reply by SimoTod** [reply]
Nah, this project does not use the issue tab, Caleb will close it (https://twitter.com/calebporzio/status/1430564014464159749).

~It's something about the scope because using an arrow function does not trigger the error (you might want to use this approach for now).~ Ignore this line, it wouldn't initialise your array

**Comment by GamesmenJordan** [comment]
Confirmed this is still an issue with 3.3.5

**Comment by thephoton** [comment]
I came across the same issue today - all tested 3.x versions of Alpine JS seem to be affected:
https://jsfiddle.net/kc5r2m0y/ - v3, throws error in DevTools console
https://jsfiddle.net/kc5r2m0y/1/ - v2, same code, no errors

In my case the initial rendering works, as does the model binding:
- click 'Toggle'
- tick a box
- 'Print Data'
- the data is correctly changed
- click 'Toggle' -> error is thrown

I'm getting around it for now by saving each property (luckily there's only 2 that change in the model) to a separate array, then watching that array for changes.
